storm-core/src/jvm/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -0,0 +1,252 @@
+                return blacklistHost.size();
+            }
+        });
+    }
+
+    @Override
+    public void schedule(Topologies topologies, Cluster cluster) {
+        LOG.debug("running Black List scheduler");
+        Map<String, SupervisorDetails> supervisors = cluster.getSupervisors();
+        for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
+            SupervisorDetails supervisorDetails = entry.getValue();
+            String hosts = supervisorDetails.getHost();
+            Set<Integer> ports = supervisorDetails.getAllPorts();
+            LOG.debug("supervisor: {}, ports: {}", hosts, ports);
+        }
+        LOG.debug("AssignableSlots: {}", cluster.getAssignableSlots());
+        LOG.debug("AvailableSlots: {}", cluster.getAvailableSlots());
+        LOG.debug("UsedSlots: {}", cluster.getUsedSlots());
+
+        blacklistStrategy.resumeFromBlacklist();
+        badSupervisors(supervisors);
+        cluster.setBlacklistedHosts(getBlacklistHosts(cluster, topologies));
+        removeLongTimeDisappearFromCache();
+
+        underlyingScheduler.schedule(topologies, cluster);
+    }
+
+    private void badSupervisors(Map<String, SupervisorDetails> supervisors) {
+        Set<String> cachedSupervisorsKeySet = cachedSupervisors.keySet();
+        Set<String> supervisorsKeySet = supervisors.keySet();
+
+        Set<String> badSupervisorKeys = Sets.difference(cachedSupervisorsKeySet, supervisorsKeySet);//cached supervisor doesn't show up
+        HashMap<String, Set<Integer>> badSupervisors = new HashMap<String, Set<Integer>>();
+        for (String key : badSupervisorKeys) {
+            badSupervisors.put(key, cachedSupervisors.get(key));
+        }
+
+        for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
+            String key = entry.getKey();
+            SupervisorDetails supervisorDetails = entry.getValue();
+            if (cachedSupervisors.containsKey(key)) {
+                Set<Integer> badSlots = badSlots(supervisors, key);
+                if (badSlots.size() > 0) {//supervisor contains bad slots
+                    badSupervisors.put(key, badSlots);
+                }
+            } else {
+                cachedSupervisors.put(key, supervisorDetails.getAllPorts());//new supervisor to cache
+            }
+        }
+
+        badSupervisorsTolerance.add(badSupervisors);
+    }
+
+    private Set<Integer> badSlots(Map<String, SupervisorDetails> supervisors, String supervisorKey) {
+        SupervisorDetails supervisor = supervisors.get(supervisorKey);
+        Set<Integer> cachedSupervisorPorts = cachedSupervisors.get(supervisorKey);
+        Set<Integer> supervisorPorts = supervisor.getAllPorts();
+
+        Set<Integer> newPorts = Sets.difference(supervisorPorts, cachedSupervisorPorts);
+        if (newPorts.size() > 0) {
+            cachedSupervisors.put(supervisorKey, Sets.union(newPorts, supervisor.getAllPorts()));
```

Comment:
Not exactly.
Image that:
In normal case the supervisor.slots.ports:[1,2,3,4],and the cachedSupervisors is also [1,2,3,4].Now port 1 break down ,then the input supervisorPorts now will be [2,3,4], but we'd like to keep cachedSupervisors to be [1,2,3,4] not [2,3,4].
The union operation is for this case : 
we reconfigure the supervisor.slots.ports form [1,2,3,4] to [2,3,4,5](may be port 1 is used for sth. else). The cachedSupervisors will now be [1,2,3,4,5] now. and port 1 will be added to blacklist soon because it will always be badSlot. But after the **BLACKLIST_SCHEDULER_TOLERANCE_TIME** ,it will be removed from cachedSupervisors  by **removeLongTimeDisappearFromCache()**, and the supervisor will be resumed from blacklist.
In this case, the supervisor which changed slot configure will be miss-added to blacklist for **BLACKLIST_SCHEDULER_RESUME_TIME**. But in the case that this will occasionally, it is acceptable.